src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+        {
+            return Subscribe(observer, null);
+        }
+
+        /// <summary>
+        /// Make a new TelemetryListener, it is a NotificationSource, which means the returned result can be used to 
+        /// log notifications, but it also has a Subscribe method so notifications can be forwarded
+        /// arbitrarily.  Thus its job is to forward things from the producer to all the listeners
+        /// (multi-casting).    Generally you should not be making your own TelemetryListener but use the
+        /// TelemetryListener.Default, so that notifications are as 'public' as possible.  
+        /// </summary>
+        public TelemetryListener(string name)
+        {
+            Name = name;
+            // To avoid allocating an explicit lock object I lock the Default TelemetryListener.   However there is a 
+            // chicken and egg problem because I need to call this code to initialize Default.   
+            var lockObj = DefaultListener;
+            if (lockObj == null)
+                lockObj = this;
+
+            // Insert myself into the list of all Listeners.   
+            lock (lockObj)
+            {
+                // Issue the callback for this new telemetry listener. 
+                var callback = s_allListenersCallback;
+                if (callback != null)
+                    callback(this);
+
+                // And add it to the list of all past listeners.  
+                m_next = s_allListeners;
+                s_allListeners = this;
+            }
+        }
+
+        /// <summary>
+        /// Clean up the NotificationListeners.   Notification listeners do NOT DIE ON THEIR OWN
+        /// because they are in a global list (for discoverability).  You must dispose them explicitly. 
+        /// Note that we do not do the Dispose(bool) pattern because we frankly don't want to support
+        /// subclasses that have non-managed state.   
+        /// </summary>
+        virtual public void Dispose()
+        {
+            // Remove myself from the list if all listeners.  
+            lock (DefaultListener)
+            {
+                m_disposed = true;
+                if (s_allListeners == this)
+                    s_allListeners = s_allListeners.m_next;
+                else
+                {
+                    var cur = s_allListeners;
+                    while (cur != null)
+                    {
+                        if (cur.m_next == this)
+                        {
+                            cur.m_next = this.m_next;
+                            break;
+                        }
+                        cur = cur.m_next;
+                    }
+                }
+                m_next = null;
+            }
+
+            // TODO audit for races...  Should this be under the lock above?  
+
+            // Indicate completion to all subscribers.  
+            Subscription subscriber = null;
+            Interlocked.Exchange(ref subscriber, m_subscriptions);
```

Comment:
The worry was two threads racing in 'Dispose' would not both send 'OnCompleted' messages.     This code was written before the _disposed variable was added, so arguably this can't happen.    Still we do want to null out the _subscriptions variable and we do want current subscriptions and doing it atomically is what you want semantically, so the Interlocked.Exchange it really not bad even if other conditions insure bad races don't happen.